### DISCUSSION TITLE
Add the ability to let project-config/apply uninstall plugins only

### DISCUSSION THREAD

**Discussion by Wiejeben** [craftcms/cms#8940] [Ideas]
### Description

When deploying I have to take in mind that maybe a plugin was removed. Normally when deploying I was first executing `composer install` and after that I was executing `./craft project-config/apply`. This would cause the "migrate down" to not be executed from the deleted plugin because Composer uninstalled the plugin(s) already.

To fix this I was thinking a variant of `project-config/apply` would solve this. This variant would exclusively execute the uninstalling of plugins, so a "pre-dependency-removal project config application". Right now I'm solving this by adding the `silent-exit-on-exception` flag to the apply execution before running `composer install`.

**Comment by brandonkelly** [comment]
> Right now I'm solving this by adding the `silent-exit-on-exception` flag to the apply execution before running `composer install`.

Are you using that on the `project-config/apply` command? If so, I would strongly discourage you from running `project-config/apply` before running `composer install`. That will lead to other problems for you down the road.

For now, a better workflow would be to keep the plugin Composer-installed until it has been DB-uninstalled across all environments.

I’ve been thinking it might be a good idea to drop Craft down to a single migration track  at some point, with all Craft/plugin/module migrations getting copied into the `migrations/` folder before executing. That would solve this as well, so long as plugins’ uninstall migrations are completely self-contained, and not relying on the plugin files to be Composer-installed.

**Comment by Wiejeben** [comment]
Yeah right now I am silencing the `apply` command. It is very rare to uninstall a plugin anyway. I just removed the apply before `composer/install` and will be deploying twice when a plugin is to be uninstalled from now on (of course until a more permanent solution is found)!

**Comment by Wiejeben** [comment]
Would be cool if this enhancement would create a system that could maybe make use of the automatic composer events. `pre-install-cmd` to uninstall plugins that are about to be uninstalled and `post-install-cmd` to apply all project config changes.
